src/lib/pipe_lib/broadcast_pipe.ml

```diff
@@ -94,13 +84,122 @@ module Writer = struct
+          | Leaf writer -> Pipe.close writer
+          | Node {child_pipe; _} -> Option.iter (Ivar.peek child_pipe) ~f:close ) ;
         Int.Table.clear t.pipes )
 end
 
+let create a =
+  let root_r, root_w = Pipe.create () in
+  let t =
+    {root_pipe= root_w; cache= a; reader_id= 0; pipes= Int.Table.create ()}
+  in
+  let downstream_flushed_v : unit Ivar.t ref = ref @@ Ivar.create () in
+  let consumer =
+    Pipe.add_consumer root_r ~downstream_flushed:(fun () ->
+        let%map () = Ivar.read !downstream_flushed_v in
+        (* Sub-pipes are never closed without closing the master pipe. *)
+        `Ok )
+  in
+  don't_wait_for
+    (Pipe.iter ~flushed:(Consumer consumer) root_r ~f:(fun v ->
+         t.cache <- v ;
+         downstream_flushed_v := Ivar.create () ;
+         let inner_pipes = Int.Table.data t.pipes in
+         let%bind () =
+           Deferred.List.iter ~how:`Parallel inner_pipes
+             ~f:(function
+             | Leaf writer -> Pipe.write writer v
+             | Node {write; _} -> write v )
+         in
+         Pipe.Consumer.values_sent_downstream consumer ;
+         let%bind () =
+           Deferred.List.iter ~how:`Parallel inner_pipes
+             ~f:(function
+             | Leaf writer -> Deferred.ignore @@ Pipe.downstream_flushed writer
+             | Node _ ->
+                 Deferred.unit
+                 (* A Broadcast pipe would get it's write determined after all of it's readers are consumed *) )
+         in
+         Ivar.fill !downstream_flushed_v () ;
+         Deferred.unit )) ;
+  (t, t)
+
+module Reader = struct
+  include Reader0
+
+  (* The only reason for a descendant pipe to get removed from it's parent if it closes.
+      The only way to do that right now is to close the pipe from the parent.
+      Therefore, we do not have to be concerned about removing a mapped pipe. *)
+  let map (t : 'a t) ~(f : 'a -> 'b) =
+    guard_already_closed t (fun () ->
+        let child_reader, child_writer = create @@ f (peek t) in
+        let reader_id = fresh_reader_id t in
+        let map_write a = Writer.write child_writer @@ f a in
+        let child_pipe = Ivar.create_full child_writer in
+        Int.Table.add_exn t.pipes ~key:reader_id
+          ~data:(Node {write= map_write; child_pipe}) ;
+        child_reader )
+
+  let filter : type a. a t -> f:(a -> bool) -> a t Ivar.t =
+   fun t ~f ->
+    guard_already_closed t (fun () ->
+        let (filter_pipe_ivar : a t Ivar.t) = Ivar.create () in
+        let reader_id = fresh_reader_id t in
+        let filter_write a =
+          if f a then
+            if Ivar.is_empty filter_pipe_ivar then (
+              let pipe, _ = create a in
+              Ivar.fill filter_pipe_ivar pipe ;
+              Deferred.unit )
+            else
+              let%bind pipe = Ivar.read filter_pipe_ivar in
+              Writer.write pipe a
+          else Deferred.unit
+        in
+        Int.Table.add_exn t.pipes ~key:reader_id
```

Comment:
the filtered pipe should be seeded with the current value in the parent pipe if it passes the predicate. so add `filter_write (peek t)` here